### ISSUE TITLE
Remember audio quality selection when downloading

### ISSUE THREAD

**Issue by VaclavTrpisovsky** [TeamNewPipe/NewPipe#8243]
### Checklist

- [X] I made sure that there are *no existing issues* - [open](https://github.com/TeamNewPipe/NewPipe/issues) or [closed](https://github.com/TeamNewPipe/NewPipe/issues?q=is%3Aissue+is%3Aclosed) - which I could contribute my information to.
- [X] I'm aware that this is a request for NewPipe itself and that requests for adding a new service need to be made at [NewPipeExtractor](https://github.com/TeamNewPipe/NewPipeExtractor/issues).
- [X] I have taken the time to fill in all the required details. I understand that the feature request will be dismissed otherwise.
- [X] This issue contains only one feature request.
- [X] I have read and understood the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md).

### Feature description

When downloading a video, I am prompted with a filetype (`Video`/`Audio`/`Subtitles`) selection screen. The filetype selection is *remembered* (previous selection is chosen as default) and affects what options are presented below: video quality, audio format&quality or subtitle language. 
The video quality selection is *remembered* but audio always defaults to `128kbps m4a`. 

### Why do you want this feature?

When downloading audio, I either optimise for file size (either `48kbps m4a` or `50kbps opus`, does not matter that much) or, less often, quality (`160 kbps opus`). Neither of these is the default option. I would welcome if my last choice was remembered for ease of use, or at least, if a more sensible selection (`50kbps opus` or `160kbps opus`) was chosen as default.

### Additional information

Low priority

**Comment by VaclavTrpisovsky** [comment]
Forgot to mention: Tested on `0.22.2` and `0.23.0 RC1`

**Comment by zeref-dragneel** [comment]
good 💯

**Comment by notaLonelyDay** [comment]
What should be remembered if previous audio quality is not available??

**Comment by VaclavTrpisovsky** [comment]
>What should be remembered if previous audio quality is not available??

Good point. I would suggest the closest one by bitrate, no matter the codec (`opus` and `aac` are similar in compression ratio anyway). Also, please don't overwrite the "previously selected" quality setting if such fallback value is used.

**Comment by opusforlife2** [comment]
> no matter the codec 

The codec matters to those who wish to use royalty-free codecs only, instead of patent-encumbered ones.

**Comment by VaclavTrpisovsky** [comment]
> **opusforlife2** [commented](https://github.com/TeamNewPipe/NewPipe/issues/8243#issuecomment-1119050267)

Well, I get this but

- YouTube is not a free service and you don't really get to choose which codecs are used (assuming you use NewPipe for YT)
- YT audio downloads are available in no formats other than `opus` and `m4a`, and if `160kbps opus` is unavailable, you don't get `50 kbps opus` either, so you don't really get a choice... in my experience. I have yet to encounter an exception to this rule.
- This is only the preselected option, and I think this strategy is better for most users (it seeks to preserve the user's choice between bandwidth reduction and quality maximisation), and certainly better than the default `128kbps m4a` every time. As much as I understand your cause, if I were to choose which users need to make 2 more taps, I would give more convenience to those who rush to get their files over public Wi-Fi or want high-quality downloads (in the realm of possibilities) rather than those who fight for codec freedom.

However, I am not the one who will be making the choice, as I am not a NewPipe developer, [thankfully](https://github.com/TeamNewPipe/NewPipe/issues/8243#issuecomment-1119050267).

**Comment by opusforlife2** [comment]
> you don't really get to choose which codecs are used 

True. There's an issue open for that. (I'm assuming you're talking about opus and m4a, because you _can_ choose between VP9 and AVC for videos).

**Comment by VaclavTrpisovsky** [comment]
>I'm assuming you're talking about opus and m4a

This thread is for audio-only downloads. I opened this issue because the audio download lacked a feature that video download already had. I don't have much experience with video downloads but if you want improvement for those, please create a new thread and maybe let me know. 

For audio, YouTube only "offers" `opus` and `m4a` AFAIK, and if `160kbps opus` is not available, `50kbps opus` is not either (in my experience, feel free to prove me wrong). So you still need to use `m4a` as the fallback format, and therefore [my proposed strategy](https://github.com/TeamNewPipe/NewPipe/issues/8243#issuecomment-1117885007) makes sense even in your use case.

----

By the way, has a task list, timeline or assignment for this enhancement been made? I am not very familiar with Github so I'm trying to apply Bugzilla logic... I'm not in a hurry but would like to know if this issue is going to be addressed along with #8035, later or TBD.

**Comment by opusforlife2** [comment]
> By the way, has a task list, timeline or assignment for this enhancement been made? I am not very familiar with Github so I'm trying to apply Bugzilla logic... I'm not in a hurry but would like to know if this issue is going to be addressed along with https://github.com/TeamNewPipe/NewPipe/issues/8035, later or TBD.

@VaclavTrpisovsky No, Newpipe has a small volunteer team. People work on what they have time for, and sometimes contributors take up issues they like as well. So there is no timeline here. If you're lucky, someone will take this up soon.

**Comment by notaLonelyDay** [comment]
Hi, @mhmdanas I would love to work on this issue. Please assign it to me.

**Comment by notaLonelyDay** [comment]
@VaclavTrpisovsky @opusforlife2 
what do you think about this flow

for every quality save last used time
and then when downloading just sort available qualities by last used time and select the most last used??

Maybe it is better than by bitrate?
Because to my mind usually you need requested feature to download a lot similar videos and they have the same qualities available

**Comment by opusforlife2** [comment]
> The video quality selection is remembered but audio always defaults to 128kbps m4a.

@VaclavTrpisovsky Actually, now that I think about it, that happens because you have M4A as your default audio format in settings. If you change that to Opus, that would solve your issue as well, by always having 160 kbps Opus pre-selected.

> or every quality save last used time
and then when downloading just sort available qualities by last used time and select the most last used?

@notaLonelyDay Sounds okay to me. 👍

**Comment by obar** [comment]
I found this issue via the search; happy to see this minor UI hiccup I was about to report is already being discussed.

> that happens because you have M4A as your default audio format in settings. If you change that to Opus, that would solve your issue as well, by always having 160 kbps Opus pre-selected.

This worked for me (though "Opus" wasn't an option in that menu, it was WebM) and as you point out it defaults to 160kbps without a settable preference. I do think the "default to the last intentionally selected mode" behavior proposed here might be more intuitive and doesn't require changing the playback mode. It already seems to work this way for the video/audio/captions selection.

It's an interesting point about the behavior when the preferred format isn't available. I can see benefits in either way and I expect it'll be a rare case. Either way, it could be useful in case to draw attention to it. Only when this happens, perhaps show the following just below the format selection: 

**Remember this as the default download format for next time**  {switch, default off} 
Your previous default, {Opus 70kbps}, isn't available for this download

**Comment by emailbin** [comment]
Would there be a way to always automatically select the smallest quality, and thus also the smallest file size? I use the audio downloader to download only spoken material and then use it as an audio book for long journeys or walks.

**Comment by VaclavTrpisovsky** [comment]
> Would there be a way to always automatically select the smallest quality, and thus also the smallest file size? [...]

These are not the same. About ¼ to ⅓ of videos are smaller as `50kbps opus` than `48kbps m4a` for some reason. My suggested implementation would solve this, though.

**Comment by daakadence** [comment]
@notaLonelyDay Can I bump this request in case you've found the time? I think it would add value to users who, like me, have limited data plans and prefer LQ downloads on the go. I usually DL ahead of time, but also prefer smaller files for the space (I mostly listen to comedy so the fidelity is not an issue). If we could opt out change to _last set_ with a fallback to m4a-128 that would be awesome. Thanks

**Comment by axu2** [comment]
Indeed, I always download the smallest opus and would love it to remember instead of defaulting to 160k opus. 

I play the audio using AntennaPod using the local folder podcast feature.
